x/cdp/keeper/querier.go

```diff
@@ -178,3 +183,85 @@ func queryGetAccounts(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) ([]
 	}
 	return bz, nil
 }
+
+// query get savings rate distributed in the cdp store
+func queryGetSavingsRateDistributed(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) ([]byte, error) {
+	// Get savings rate distributed
+	savingsRateDist := keeper.GetSavingsRateDistributed(ctx)
+
+	// Encode results
+	bz, err := codec.MarshalJSONIndent(types.ModuleCdc, savingsRateDist)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrJSONMarshal, err.Error())
+	}
+	return bz, nil
+}
+
+// query cdps in store and filter by request params
+func queryGetV2Cdps(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) ([]byte, error) {
+	var params types.QueryV2CdpsParams
+	err := types.ModuleCdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrJSONUnmarshal, err.Error())
+	}
+
+	// Filter CDPs
+	unfilteredCDPs := keeper.GetAllCdps(ctx)
```

Comment:
Putting all CDPs in memory and filtering is too inefficient as a default querying strategy.

* Check the query params supplied and apply the best keeper function. For example, if collateral type and ID is supplied, call `GetCDP`, if collateral type only is supplied, call `GetCDPsByCollateralType`. 
* I think we already have most of the methods we want, except for `GetCDPsByOwner` (ie, Owner is supplied in the query params, but not collateral type). Since I don't really want to re-write the store at this stage, we can add that method via calling `GetCdpIdsByOwner` then brute forcing it by calling `GetCDP` and trying each denom prefix.

Reply:
Looks like `keeper_bench_test.go` can likely be used to get some tangible efficiency metrics. This in-memory filtering strategy is something I picked up from the [gov module](https://github.com/cosmos/cosmos-sdk/blob/master/x/gov/client/rest/query.go), which probably has a lighter store than our CDP module.

We're also using filtering in the `auction/auctions` and `bep3/swaps` queries - if this method proves too inefficient we should consider patching bep3 as well. Auction is probably fine, but there can easily be as many swaps in memory as cdps given that we store swaps in longterm storage before deletion.

Reply:
+1 to investigating bep3 filter method - wouldn't be surprised if we were storing thousands of swaps eventually - wouldn't want to load them all either. Agree that auctions is probably not a concern at the moment.

Reply:
Halved the memory usage by converting to AugmentedCDP in the filter loop (as a byproduct of ratio param implementation requirements), but that's still only a constant O(n) improvement. So still need to address this one.

Reply:
@karzak added the proposed CDP query result filtering in https://github.com/Kava-Labs/kava/pull/644/commits/1c6d5d52ed5f6505b9bc81c8e05943090a6ca8ac. The BEP3 module doesn't have any of the analogous keeper methods, so updating the querier would require significant store modifications. The query filtering is still done in linear time.